index.js

```diff
@@ -1,469 +1,173 @@
-const vm = require("vm");
+const BrowserTab = require("./lib/BrowserTab");
 const {version} = require("./package.json");
-const {CookieAccessInfo, Cookie} = require("cookiejar");
-const {Document, Window} = require("./lib");
+const {CookieAccessInfo} = require("cookiejar");
 const {normalizeHeaders, getLocationHost} = require("./lib/getHeaders");
+const saveToJar = require("./lib/saveToJar");
 
 module.exports = Tallahassee;
 
-function Tallahassee(app, options = {}) {
-  const agent = supertest.agent(app);
-  const defaultHeaders = normalizeHeaders(options.headers);
-  return {
-    jar: agent.jar,
-    navigateTo,
-  };
+const responseSymbol = Symbol.for("response");
 
-  function navigateTo(linkUrl, headers = {}, statusCode = 200) {
-    const requestHeaders = {
-      ...defaultHeaders,
-      ...normalizeHeaders(headers),
-    };
-
-    if (requestHeaders["set-cookie"]) {
-      const setCookies = requestHeaders["set-cookie"];
-      saveToJar(setCookies);
-      requestHeaders["set-cookie"] = undefined;
-    }
-
-    const webPage = WebPage(agent, requestHeaders, agent.jar);
-    return webPage.load(linkUrl, requestHeaders, statusCode);
+class OriginResponse {
+  constructor(uri, response, originHost, protocol) {
+    this[responseSymbol] = response;
+    const status = this.status = response.statusCode;
+    this.ok = status >= 200 && status < 300;
+    this.headers = new Map(Object.entries(response.headers));
+    this.url = originHost ? `${protocol}//${originHost}${uri}` : response.request.url;
   }
+  text() {
+    return Promise.resolve(this[responseSymbol].text);
+  }
+  json() {
+    return Promise.resolve(this[responseSymbol].body);
+  }
+}
 
-  function WebPage(origin, originRequestHeaders, jar) {
-    const originHost = getLocationHost(originRequestHeaders);
-    const protocol = `${originRequestHeaders["x-forwarded-proto"] || "http"}:`;
-
-    let userAgent = `Tallahassee/${version}`;
-    const referrer = originRequestHeaders.referer;
-    const page = {
-      originHost,
-      load,
-      fetch,
-    };
-
-    return page;
-
-    function load(uri, headers, statusCode = 200) {
-      const requestHeaders = normalizeHeaders(headers);
-      if (requestHeaders["user-agent"]) userAgent = requestHeaders["user-agent"];
-
-      if (requestHeaders.cookie) {
-        const publicHost = getLocationHost(requestHeaders);
-        const parsedUri = url.parse(uri);
-        const cookieDomain = parsedUri.hostname || publicHost || originHost || "127.0.0.1";
-        const isSecure = (parsedUri.protocol || protocol) === "https:";
+class WebPage {
+  constructor(agent, originRequestHeaders) {
+    this.agent = agent;
+    this.jar = agent.jar;
+    this.originRequestHeaders = originRequestHeaders;
+    this.originHost = getLocationHost(originRequestHeaders);
+    this.userAgent = `Tallahassee/${version}`;
+    this.protocol = `${originRequestHeaders["x-forwarded-proto"] || "http"}:`;
+    this.referrer = originRequestHeaders.referer;
```

Comment:
typo? referrer !== referer

Reply:
Yes, but the header is misspelled from [day 1](https://en.wikipedia.org/wiki/HTTP_referer).

Reply:
![](https://c.tenor.com/sBJAWaF-FsUAAAAM/smile-nope.gif)